File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_matlab_labs_and_solutions/command/memento_solution/SimpleDriverCommand.m
% SimpleDriverCommand.m

classdef (Abstract) SimpleDriverCommand < Command
	properties (GetAccess = protected, SetAccess = private)
		Driver % DpDriver
	end

	properties (GetAccess = private, SetAccess = private)
		DpDriverMementoInterface % DpDriverMementoInterface
	end

	methods
		% constructor:
		function theSimpleDriverCommand = SimpleDriverCommand(aDriver)
			theSimpleDriverCommand.Driver = aDriver;
		end

		function execute(theSimpleDriverCommand)
			theSimpleDriverCommand.DpDriverMementoInterface = theSimpleDriverCommand.Driver.backup;
		end

		function undo(theSimpleDriverCommand)
			theSimpleDriverCommand.Driver.restore(theSimpleDriverCommand.DpDriverMementoInterface);
		end

		function set.Driver(theSimpleDriverCommand, aDriver)
			if isempty(aDriver)
				error('aDriver is empty.');
			end

			theSimpleDriverCommand.Driver = aDriver;
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/command/memento_solution/SimpleDriverCommand.java
// SimpleDriverCommand.java

public abstract class SimpleDriverCommand extends Command
{
	public SimpleDriverCommand(Driver aDriver)
	{
		setDriver(aDriver);
		myDriverMementoInterface = null;
	}

	public void execute()
	{
		myDriverMementoInterface = getDriver().backup();
	}

	public void undo()
	{
		getDriver().restore(myDriverMementoInterface);
	}

	private void setDriver(Driver aDriver)
	{
		if (aDriver == null)
		{
			throw new DpException("void SimpleDriverCommand.setDriver(Driver aDriver) -- aDriver == null.");
		}

		myDriver = aDriver;
	}

	protected Driver getDriver()
	{
		return myDriver;
	}

	private Driver myDriver;
	private Driver.MementoInterface myDriverMementoInterface;
}
